From: cl2t <yihong.or@isrc.iscas.ac.cn>
Subject: [PATCH] Order account locks when revalidating reciprocal imports

Parallel claim updates for accounts with reciprocal imports can each hold
an importing account's write lock while waiting for the exporting account's
read lock. This blocks account updates and client reconnects indefinitely.

Acquire both account locks in name order for stream and service import
revalidation, then use the authorization helpers that expect the exporter
lock to be held. Service export metadata is also read under that lock.
Add concurrent reciprocal stream and service claim update regressions.

Signed-off-by: cl2t <yihong.or@isrc.iscas.ac.cn>
---
--- a/server/accounts.go
+++ b/server/accounts.go
@@ -3281,6 +3281,23 @@
 	return false
 }
 
+// Lock distinct exporting and importing accounts in name order to avoid a
+// cycle when accounts with reciprocal imports refresh their claims together.
+func (a *Account) lockForImportValidation(importer *Account) {
+	if a.Name < importer.Name {
+		a.mu.RLock()
+		importer.mu.Lock()
+	} else {
+		importer.mu.Lock()
+		a.mu.RLock()
+	}
+}
+
+func (a *Account) unlockForImportValidation(importer *Account) {
+	a.mu.RUnlock()
+	importer.mu.Unlock()
+}
+
 // updateAccountClaimsWithRefresh will update an existing account with new claims.
 // If refreshImportingAccounts is true it will also update incomplete dependent accounts
 // This will replace any exports or imports previously defined.
@@ -3552,21 +3569,18 @@
 			if acc.Name == a.Name {
 				return true
 			}
-			// TODO: checkStreamImportAuthorized() stack should not be trying
-			// to lock "acc". If we find that to be needed, we will need to
-			// rework this to ensure we don't lock acc.
-			acc.mu.Lock()
+			a.lockForImportValidation(acc)
 			for _, im := range acc.imports.streams {
 				if im != nil && im.acc.Name == a.Name {
 					// Check for if we are still authorized for an import.
-					im.invalid = !a.checkStreamImportAuthorized(acc, im.from, im.claim)
+					im.invalid = !a.checkStreamImportAuthorizedNoLock(acc, im.from, im.claim)
 					awcsti[acc.Name] = struct{}{}
 					for c := range acc.clients {
 						clients[c] = struct{}{}
 					}
 				}
 			}
-			acc.mu.Unlock()
+			a.unlockForImportValidation(acc)
 			return true
 		})
 		// Now walk clients.
@@ -3582,14 +3596,11 @@
 			if acc.Name == a.Name {
 				return true
 			}
-			// TODO: checkServiceImportAuthorized() stack should not be trying
-			// to lock "acc". If we find that to be needed, we will need to
-			// rework this to ensure we don't lock acc.
-			acc.mu.Lock()
+			a.lockForImportValidation(acc)
 			for _, si := range acc.imports.services {
 				if si != nil && si.acc.Name == a.Name {
 					// Check for if we are still authorized for an import.
-					si.invalid = !a.checkServiceImportAuthorized(acc, si.to, si.claim)
+					si.invalid = !a.checkServiceImportAuthorizedNoLock(acc, si.to, si.claim)
 					// Make sure we should still be tracking latency and if we
 					// are allowed to trace.
 					if !si.response {
@@ -3603,7 +3614,7 @@
 					}
 				}
 			}
-			acc.mu.Unlock()
+			a.unlockForImportValidation(acc)
 			return true
 		})
 	}
--- a/server/jwt_test.go
+++ b/server/jwt_test.go
@@ -7172,3 +7172,55 @@
 	require_True(t, exporter.checkStreamImportAuthorized(importer, "foo.one", nil))
 	require_True(t, exporter.checkStreamImportAuthorized(importer, "bar.one", nil))
 }
+
+func TestJWTConcurrentReciprocalClaims(t *testing.T) {
+	for _, kind := range []jwt.ExportType{jwt.Stream, jwt.Service} {
+		t.Run(kind.String(), func(t *testing.T) {
+			opts := DefaultOptions()
+			opts.NoSystemAccount = true
+			s, err := NewServer(opts)
+			require_NoError(t, err)
+			a, err := s.RegisterAccount("accountA")
+			require_NoError(t, err)
+			b, err := s.RegisterAccount("accountB")
+			require_NoError(t, err)
+			makeClaim := func(acc *Account, extra bool) *jwt.AccountClaims {
+				claim := jwt.NewAccountClaims(acc.Name)
+				subject, other, imported := "foo", b.Name, "bar"
+				if acc == b {
+					subject, other, imported = "bar", a.Name, "foo"
+				}
+				claim.Exports.Add(&jwt.Export{Subject: jwt.Subject(subject), Type: kind})
+				if extra {
+					claim.Exports.Add(&jwt.Export{Subject: "other", Type: kind})
+				}
+				claim.Imports.Add(&jwt.Import{Account: other, Subject: jwt.Subject(imported), Type: kind})
+				return claim
+			}
+			ca, cb := makeClaim(a, false), makeClaim(b, false)
+			ca.Imports = nil
+			cb.Imports = nil
+			s.UpdateAccountClaims(a, ca)
+			s.UpdateAccountClaims(b, cb)
+			s.UpdateAccountClaims(a, makeClaim(a, false))
+			s.UpdateAccountClaims(b, makeClaim(b, false))
+			var wg sync.WaitGroup
+			wg.Add(2)
+			for _, acc := range []*Account{a, b} {
+				go func(acc *Account) {
+					defer wg.Done()
+					for i := 0; i < 1000; i++ {
+						s.UpdateAccountClaims(acc, makeClaim(acc, i%2 == 0))
+					}
+				}(acc)
+			}
+			done := make(chan struct{})
+			go func() { wg.Wait(); close(done) }()
+			select {
+			case <-done:
+			case <-time.After(10 * time.Second):
+				t.Fatal("reciprocal account claim updates did not complete")
+			}
+		})
+	}
+}
